File: app/Models/Ubicacion.php
<?php
declare(strict_types=1);

class Ubicacion
{
    private PDO $db;

    public function __construct()
    {
        $this->db = Database::getInstance();
    }

    public function all(): array
    {
        $stmt = $this->db->query(
            "SELECT u.*, us.nombre AS usuario_nombre, us.usuario AS usuario_user
             FROM ubicaciones u
             JOIN usuarios us ON u.user_id = us.id
             WHERE u.deleted_at IS NULL
             ORDER BY u.created_at DESC"
        );
        return $stmt->fetchAll();
    }

    public function find(int $id): ?array
    {
        $stmt = $this->db->prepare(
            "SELECT u.*, us.nombre AS usuario_nombre, us.usuario AS usuario_user
             FROM ubicaciones u
             JOIN usuarios us ON u.user_id = us.id
             WHERE u.id = :id AND u.deleted_at IS NULL LIMIT 1"
        );
        $stmt->execute(['id' => $id]);
        return $stmt->fetch() ?: null;
    }

    public function byUsuario(int $userId): array
    {
        $stmt = $this->db->prepare(
            "SELECT * FROM ubicaciones WHERE user_id = :user_id AND deleted_at IS NULL ORDER BY created_at DESC"
        );
        $stmt->execute(['user_id' => $userId]);
        return $stmt->fetchAll();
    }

    public function create(array $data): int
    {
        $stmt = $this->db->prepare(
            "INSERT INTO ubicaciones (lat, lng, nombre, descripcion, user_id)
             VALUES (:lat, :lng, :nombre, :descripcion, :user_id)"
        );
        $stmt->execute([
            'lat'         => $data['lat'],
            'lng'         => $data['lng'],
            'nombre'      => $data['nombre'],
            'descripcion' => $data['descripcion'] ?? null,
            'user_id'     => $data['user_id'],
        ]);
        return (int) $this->db->lastInsertId();
    }

    public function update(int $id, array $data): bool
    {
        $stmt = $this->db->prepare(
            "UPDATE ubicaciones
             SET lat = :lat, lng = :lng, nombre = :nombre, descripcion = :descripcion
             WHERE id = :id AND deleted_at IS NULL"
        );
        return $stmt->execute([
            'lat'         => $data['lat'],
            'lng'         => $data['lng'],
            'nombre'      => $data['nombre'],
            'descripcion' => $data['descripcion'] ?? null,
            'id'          => $id,
        ]);
    }

    public function delete(int $id): bool
    {
        $stmt = $this->db->prepare(
            "UPDATE ubicaciones SET deleted_at = NOW() WHERE id = :id AND deleted_at IS NULL"
        );
        return $stmt->execute(['id' => $id]);
    }
}

File: aplicacion/lib/models/ubicacion.dart
class Ubicacion {
  final int id;
  final double lat;
  final double lng;
  final String nombre;
  final String? descripcion;
  final String? url;
  final int userId;
  final String? usuarioNombre;

  const Ubicacion({
    required this.id,
    required this.lat,
    required this.lng,
    required this.nombre,
    this.descripcion,
    this.url,
    required this.userId,
    this.usuarioNombre,
  });

  factory Ubicacion.fromJson(Map<String, dynamic> j) => Ubicacion(
        id: j['id'] is int ? j['id'] : int.parse(j['id'].toString()),
        lat: double.parse(j['lat'].toString()),
        lng: double.parse(j['lng'].toString()),
        nombre: j['nombre'] ?? '',
        descripcion: j['descripcion'],
        url: j['url'],
        userId: j['user_id'] is int
            ? j['user_id']
            : int.parse(j['user_id'].toString()),
        usuarioNombre: j['usuario_nombre'],
      );
}

class Usuario {
  final int id;
  final String nombre;
  final String usuario;
  final String? cargo;
  final String rol;

  const Usuario({
    required this.id,
    required this.nombre,
    required this.usuario,
    this.cargo,
    required this.rol,
  });

  factory Usuario.fromJson(Map<String, dynamic> j) => Usuario(
        id: j['id'] is int ? j['id'] : int.parse(j['id'].toString()),
        nombre: j['nombre'] ?? '',
        usuario: j['usuario'] ?? '',
        cargo: j['cargo'],
        rol: j['rol'] ?? 'user',
      );

  Map<String, dynamic> toJson() => {
        'id': id,
        'nombre': nombre,
        'usuario': usuario,
        'cargo': cargo,
        'rol': rol,
      };
}
